spec/integration/secure_uploads_spec.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+describe "Secure uploads" do
+  fab!(:user) { Fabricate(:user) }
+  fab!(:group) { Fabricate(:group) }
+  fab!(:secure_category) { Fabricate(:private_category, group: group) }
+
+  before do
+    Jobs.run_immediately!
+
+    # this is done so the after_save callbacks for site settings to make
+    # UploadReference records works
+    @original_provider = SiteSetting.provider
+    SiteSetting.provider = SiteSettings::DbProvider.new(SiteSetting)
+    setup_s3
+    stub_s3_store
+    SiteSetting.secure_uploads = true
+    group.add(user)
+    user.reload
+  end
+
+  after { SiteSetting.provider = @original_provider }
+
+  def create_upload
+    filename = "logo.png"
+    file = file_from_fixtures(filename)
+    UploadCreator.new(file, filename).create_for(user.id)
+  end
+
+  def stub_presign_upload_get(upload)
+    # this is necessary because by default any upload inside a secure post is considered "secure"
+    # for the purposes of fetching hotlinked images until proven otherwise, and this is easier
+    # than trying to stub the presigned URL for s3 in a different way
+    stub_request(:get, "https:#{upload.url}").to_return(
+      status: 200,
+      body: file_from_fixtures("logo.png"),
+    )
+    Upload.stubs(:signed_url_from_secure_uploads_url).returns("https:#{upload.url}")
+  end
+
+  it "does not convert an upload to secure when it was first used in a site setting then in a post" do
+    upload = create_upload
+    SiteSetting.favicon = upload
+    expect(upload.reload.upload_references.count).to eq(1)
+    create_post(
+      title: "Secure upload post",
+      raw: "This is a new post <img src=\"#{upload.url}\" />",
+      category: secure_category,
+      user: user,
+    )
+    upload.reload
+    expect(upload.upload_references.count).to eq(2)
+    expect(upload.secure).to eq(false)
+  end
+
+  it "does not convert an upload to insecure when it was first used in a secure post then a site setting" do
```

Comment:
In this case, wouldn't it mean that the same problem will still occur? I kind of feel like if an upload has been used for a site setting, it means we can safely mark it as not secured even if it was a secured upload before.

Reply:
No because then this is a kind of "security issue" in that you could have some super secret image that is uploaded to a site setting and then made public...far fetched I know but we did discuss this in the original designs of secure uploads. We want the first place that the thing was used to determine security, it makes things a lot easier to reason about.

At some point we discussed preventing people from doing things like this e.g. throwing some error if an upload from a private post was posted in a public topic, but we never went down that road. Maybe someday, I am not sure.